File: leetcode/problems/3.Longest-Substring-Without-Repeating-Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.size() == 0) 
            return 0; 
            
        queue<char> q; 
        unordered_map<char, int> mp; 
        int ans = 1; 
        for (int i = 0; i < s.size(); ++i) {
            char c = s[i]; 
            while (mp[c] > 0) {
                --mp[q.front()];
                q.pop(); 
            }
            q.push(c); 
            ++mp[c]; 
            if (q.size() > ans)
                ans = q.size(); 
        }
        return ans; 
    }
};

File: leetcode/study-plans/Algorithms-I/day6/3.Longest-Substring-Without-Repeating-Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        q = []
        reps = {}
        ans = 0
        for c in s: 
            q.append(c)
            if c in reps.keys(): 
                reps[c] += 1 
            else: 
                reps[c] = 1 
            
            while reps[c] > 1: 
                reps[q[0]] -= 1 
                q.pop(0)
                
            ans = max(ans, len(q)) 
            
        return ans 
        
